python/instrumentation/openinference-instrumentation-litellm/src/openinference/instrumentation/litellm/__init__.py

```diff
@@ -0,0 +1,182 @@
+        )
+        self._set_span_attribute(span, SpanAttributes.LLM_MODEL_NAME, kwargs.get("model"))
+        self._set_span_attribute(span, SpanAttributes.INPUT_VALUE, str(kwargs.get("prompt")))
+
+    def _finalize_span(self, span: trace.Span, result: Any) -> None:
+        if isinstance(result, litellm.ModelResponse):
+            self._set_span_attribute(
+                span, SpanAttributes.OUTPUT_VALUE, result.choices[0].message.content
+            )
+        elif isinstance(result, litellm.EmbeddingResponse):
+            if len(result.data) > 0:
+                first_embedding = result.data[0]
+                self._set_span_attribute(
+                    span,
+                    EmbeddingAttributes.EMBEDDING_VECTOR,
+                    json.dumps(first_embedding.get("embedding", [])),
+                )
+        elif isinstance(result, litellm.ImageResponse):
+            if len(result.data) > 0:
+                self._set_span_attribute(span, ImageAttributes.IMAGE_URL, result.data[0]["url"])
+                self._set_span_attribute(span, SpanAttributes.OUTPUT_VALUE, result.data[0]["url"])
+        if hasattr(result, "usage"):
+            self._set_span_attribute(
+                span, SpanAttributes.LLM_TOKEN_COUNT_PROMPT, result.usage["prompt_tokens"]
+            )
+            self._set_span_attribute(
+                span, SpanAttributes.LLM_TOKEN_COUNT_COMPLETION, result.usage["completion_tokens"]
+            )
+            self._set_span_attribute(
+                span, SpanAttributes.LLM_TOKEN_COUNT_TOTAL, result.usage["total_tokens"]
+            )
+
+    def _instrument(self, tracer_provider: Optional[TracerProvider] = None) -> None:
+        if tracer_provider:
+            self.tracer_provider = tracer_provider
+            trace.set_tracer_provider(tracer_provider)
+        self.tracer = trace.get_tracer(__name__)
+
+        functions_to_instrument = {
+            "completion": "completion",
+            "acompletion": "acompletion",
+            "completion_with_retries": "completion_with_retries",
+            # Bug report filed on GitHub for acompletion_with_retries: https://github.com/BerriAI/litellm/issues/4908
+            # 'acompletion_with_retries': 'acompletion_with_retries',
+            "embedding": "embedding",
+            "aembedding": "aembedding",
+            "image_generation": "image_generation",
+            "aimage_generation": "aimage_generation",
+        }
+
+        for func_name, span_name in functions_to_instrument.items():
+            if hasattr(litellm, func_name):
+                original_func = getattr(litellm, func_name)
+                LiteLLMInstrumentor.original_litellm_funcs[func_name] = (
+                    original_func  # Add original liteLLM function to dictionary
+                )
+                setattr(litellm, func_name, self._common_wrapper(original_func, span_name))
```

Comment:
IMO this feels a bit overabstracted, since "common wrapper" as a name does suggest that something is shared between the wrappers but it's not clear exactly what. Probably no need to change this though

Reply:
Done-- each function is now mapped to a corresponding wrapper